qpicard/src/org/qcmg/picard/SAMWriterFactory.java

```diff
@@ -0,0 +1,137 @@
+		this(header, preSort, output,tmpDir, 0, true);
+	}
+	public SAMWriterFactory(SAMFileHeader header,  boolean preSort, File output,File tmpDir, boolean createindex){
+		this(header, preSort, output,tmpDir, 0, createindex);
+	}
+	
+	public SAMWriterFactory(SAMFileHeader header,  boolean preSort, File output,File tmpDir, int ramReads){
+		this(header, preSort, output,tmpDir, ramReads, true);
+	}
+	public SAMWriterFactory(SAMFileHeader header,  boolean preSort, File output,File tmpDir, int ramReads, boolean createIndex){
+		this(header, preSort, output,tmpDir, ramReads, createIndex, false);
+	}
+	public SAMWriterFactory(SAMFileHeader header,  boolean preSort, File output,File tmpDir, int ramReads, 
+			boolean createIndex, boolean useAsyncIO){
+		this(header, preSort, output,tmpDir, ramReads, createIndex, useAsyncIO, -1, null);
+	}
+	
+	public SAMWriterFactory(SAMFileHeader header,  boolean preSort, File output, File tmpDir, int ramReads, 
+			boolean createIndex, boolean useAsyncIO, int asyncOutputBufferSize ){
+		
+		this(header, preSort, output, tmpDir, ramReads, createIndex, useAsyncIO, asyncOutputBufferSize, null);
+	}
+
+	
+	public SAMWriterFactory(SAMFileHeader header,  boolean preSort, File output,File tmpDir, int ramReads, 
+			boolean createIndex, boolean useAsyncIO, int asyncOutputBufferSize, File reference ) {
+		SAMFileWriterFactory factory = new SAMFileWriterFactory();   
+		if (ramReads > 0) {
+			htsjdk.samtools.SAMFileWriterImpl.setDefaultMaxRecordsInRam( ramReads );
+		}
+		
+		index = createIndex && header.getSortOrder().equals(SAMFileHeader.SortOrder.coordinate);
+		
+		factory.setCreateIndex(index);
+		
+		/*
+		 * set async setting
+		 */
+		factory.setUseAsyncIo(useAsyncIO);
+		if (useAsyncIO && asyncOutputBufferSize > 0) {
+			factory.setAsyncOutputBufferSize(asyncOutputBufferSize);
+		}
+		
+		if (tmpDir != null) {
+			factory.setTempDirectory(tmpDir);
+		}
+		
+		this.output = output;		
+		
+        final String filename = output.getName();
+        
+        if (BAM_TYPE.hasValidFileExtension(filename)) {
+                 writer = factory.makeBAMWriter(header, preSort, output);                       
+        } else if(CRAM_TYPE.hasValidFileExtension(filename)) {   
+				File ref = (reference == null)? Defaults.REFERENCE_FASTA: reference;            
+                writer = factory.makeCRAMWriter(header, preSort, output, ref);
+        } else {
+            if (!SAM_TYPE.hasValidFileExtension(filename)) {
+                logger.error("Unknown file extension, assuming SAM format when writing file: " + filename);               
+            }
+            writer = factory.makeSAMWriter(header, preSort, output);
+        }
+	}
+	
+	public SAMFileWriter getWriter(){		
+		return writer;
+	}
+	/**
+	 * picard creates index aln.bai for aln.bam; and aln.cram.bai for aln.cram.
+	 * Here we rename aln.bai to aln.bam.bai; and lan.cram.bai to aln.cram.crai.
+	 */
+	public void renameIndex(){
```

Comment:
I think this method should be renamed to something like `closeWriterAndRenameIndex` thus explicitly depicting what the method does. The current name is misleading in that regard.